File: conversions/dedent/elide-dedent.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-dedent
#
# This demonstrates calling the TypeScript dedent implementation
# from Ruby using Elide's polyglot capabilities.
#
# Benefits:
# - One dedent implementation shared across TypeScript and Ruby
# - Consistent string formatting across microservices
# - No Ruby heredoc workarounds needed
# - Perfect for Rails SQL queries, templates

puts "=== Ruby Calling TypeScript Dedent ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# dedent_module = Elide.require('./elide-dedent.ts')

# Example 1: ActiveRecord SQL Queries
# class User < ApplicationRecord
#   def self.active_users_sql
#     dedent_module.default(<<~SQL
#       SELECT users.*, COUNT(posts.id) as post_count
#       FROM users
#       LEFT JOIN posts ON posts.user_id = users.id
#       WHERE users.active = true
#       GROUP BY users.id
#       ORDER BY post_count DESC
#     SQL
#     )
#   end
# end

# Example 2: Email Templates in Rails
# class UserMailer < ApplicationMailer
#   def welcome_email(user)
#     body = dedent_module.default(<<~TEXT
#       Hello #{user.name}!
#
#       Welcome to our platform.
#
#       Your account details:
#       - Email: #{user.email}
#       - Username: #{user.username}
#
#       Best regards,
#       The Team
#     TEXT
#     )
#     mail(to: user.email, subject: 'Welcome!', body: body)
#   end
# end

# Example 3: ERB Template Formatting
# class TemplateHelper
#   def self.render_user_card(user)
#     template = dedent_module.default(<<~HTML
#       <div class="user-card">
#         <h2><%= user.name %></h2>
#         <p><%= user.email %></p>
#         <p>Member since: <%= user.created_at.strftime('%Y-%m-%d') %></p>
#       </div>
#     HTML
#     )
#     ERB.new(template).result(binding)
#   end
# end

# Example 4: Rake Task Documentation
# namespace :users do
#   desc dedent_module.default(<<~DESC
#     Import users from CSV file
#
#     Usage: rake users:import FILE=users.csv
#
#     The CSV should have columns: name, email, username
#   DESC
#   )
#   task :import do
#     # Import logic
#   end
# end

# Example 5: Configuration Generation
# class ConfigGenerator
#   def self.generate_redis_config(host, port)
#     dedent_module.default(<<~CONF
#       host: #{host}
#       port: #{port}
#       timeout: 5
#       db: 0
#       pool: 5
#     CONF
#     )
#   end
# end

puts "Real-world use cases:"
puts "- ActiveRecord SQL query formatting"
puts "- Email template generation"
puts "- ERB/HAML template strings"
puts "- Rake task descriptions"
puts "- Configuration file generation"
puts "- Multi-line string literals in Rails"
puts ""

puts "Example: Rails E-Commerce Platform"
puts "┌──────────────────────────────────────┐"
puts "│   Elide Dedent (TypeScript)         │"
puts "│   elide-dedent.ts                   │"
puts "└──────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Rails │  │ Sidekiq│"
puts "    │  API   │  │  App   │  │ Worker │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same clean strings everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby <<~ heredoc + JS dedent = different behavior"
puts "After: One Elide implementation = consistent formatting"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/dedent.rb"
puts "  DEDENT = Elide.require('./elide-dedent.ts')"
puts "  "
puts "  # app/models/query_builder.rb"
puts "  class QueryBuilder"
puts "    def self.complex_query"
puts "      DEDENT.default(<<~SQL"
puts "        SELECT * FROM users"
puts "        WHERE active = true"
puts "      SQL"
puts "      )"
puts "    end"
puts "  end"

File: converted/utilities/dedent/elide-dedent.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-dedent

This demonstrates calling the TypeScript dedent implementation
from Python using Elide's polyglot capabilities.

Benefits:
- One dedent implementation shared across TypeScript and Python
- Consistent string formatting across services
- No Python textwrap.dedent needed
- Perfect for SQL queries, templates, multi-line strings
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# dedent_module = require('./elide-dedent.ts')

print("=== Python Calling TypeScript Dedent ===\n")

# Example 1: SQL Query Formatting
# def get_users_query():
#     query = dedent_module.default("""
#         SELECT id, username, email
#         FROM users
#         WHERE active = true
#         ORDER BY created_at DESC
#     """)
#     return query
#
# query = get_users_query()
# print("SQL Query:")
# print(query)
# print()

# Example 2: Email Templates
# class EmailService:
#     def send_welcome_email(self, username, email):
#         body = dedent_module.default(f"""
#             Hello {username}!
#
#             Welcome to our platform. We're excited to have you.
#
#             Your registered email: {email}
#
#             Best regards,
#             The Team
#         """)
#         # Send email with body
#         return body
#
# email_service = EmailService()
# email = email_service.send_welcome_email("Alice", "alice@example.com")
# print("Email Body:")
# print(email)
# print()

# Example 3: Django Template Rendering
# from django.template import Template, Context
#
# def render_user_profile(user):
#     template_str = dedent_module.default("""
#         <div class="user-profile">
#             <h1>{{ user.username }}</h1>
#             <p>Email: {{ user.email }}</p>
#             <p>Joined: {{ user.created_at }}</p>
#         </div>
#     """)
#     template = Template(template_str)
#     context = Context({'user': user})
#     return template.render(context)

# Example 4: Configuration File Generation
# class ConfigGenerator:
#     def generate_nginx_config(self, domain, port):
#         config = dedent_module.default(f"""
#             server {{
#                 listen 80;
#                 server_name {domain};
#
#                 location / {{
#                     proxy_pass http://localhost:{port};
#                     proxy_set_header Host $host;
#                 }}
#             }}
#         """)
#         return config
#
# generator = ConfigGenerator()
# config = generator.generate_nginx_config("example.com", 3000)
# print("Nginx Config:")
# print(config)
# print()

# Example 5: Docstring Formatting
# class APIClient:
#     def fetch_data(self, endpoint):
#         """
#         Use dedent for clean docstrings
#         """
#         docstring = dedent_module.default("""
#             Fetch data from the API endpoint.
#
#             Args:
#                 endpoint: The API endpoint path
#
#             Returns:
#                 JSON response from the API
#         """)
#         return docstring

print("Real-world use cases:")
print("- SQL query formatting")
print("- Email template generation")
print("- HTML/XML string formatting")
print("- Configuration file generation")
print("- Docstring and comment formatting")
print("- Multi-line string literals")
print()

print("Example: Flask Web Application")
print("┌──────────────────────────────────────┐")
print("│   Elide Dedent (TypeScript)         │")
print("│   elide-dedent.ts                   │")
print("└──────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │ Flask  │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same clean strings everywhere!")
print()

print("Problem Solved:")
print("Before: Python textwrap + JS dedent = different edge cases")
print("After: One Elide implementation = consistent formatting")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  dedent = require('./elide-dedent.ts')")
print("  clean_str = dedent.default('''")
print("      Indented text")
print("      More text")
print("  ''')")
